### DISCUSSION TITLE
Leverage `CommunityToolkit.MVVM` Messenger

### DISCUSSION THREAD

**Discussion by TheCodeTraveler** [dotnet/maui#3797] [Ideas]
The  [CommunityToolkit.MVVM](https://www.nuget.org/packages/CommunityToolkit.Mvvm/) Messenger has been optimized to avoid any allocations and is 100x more efficient than the `Microsoft.Maui.Controls.MessagingCenter` (see **Benchmark** below).

The MVVM Community Toolkit is an open-source library in the .NET Foundation created and maintained by Microsoft Engineering in conjunction with community contributions.

More information can be found in the Microsoft Docs:
https://learn.microsoft.com/dotnet/communitytoolkit/mvvm/messenger

### Recommendation: Deprecate `MessagingCenter`

I recommend adding the following `[Obsolete]` attribute:

```cs
[Obsolete("We recommend `CommunityToolkit.MVVM.WeakReferenceMessenger`: https://www.nuget.org/packages/CommunityToolkit.Mvvm" ]
```

Adding the `[Obsolete]` attribute allows for the following scenarios:
  - Allows internal .NET MAUI components to continue to use `Microsoft.Maui.Controls MessagingCenter`
  - Allows Xamarin.Forms applications migrating to .NET MAUI to continue using .NET MAUI's `MessagingCenter` until they move to `CommunityToolkit.MVVM`'s messenger 

### Benchmark

[![image](https://user-images.githubusercontent.com/13558917/146473644-5ad16590-38db-432d-baa3-39caea130bdc.png)](https://twitter.com/SergioPedri/status/1469267425522712577?s=20)
[Benchmark Source](https://twitter.com/SergioPedri/status/1469267425522712577)

**Comment by BurkusCat** [comment]
Is there potential in the future for Maui to directly use the toolkit's messenger?

Also, would the other frameworks you benchmarked against be interested in moving to this implementation? Are there any barriers to something like that happening? (if it is possible to have the most performant messenger with most of the features needed in other frameworks that would be amazing)

**Reply by hartez** [reply]
> Is there potential in the future for Maui to directly use the toolkit's messenger?

Yep. This is what that might look like: https://github.com/dotnet/maui/pull/3880
